hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/index/record/RecordIndexer.java

```diff
@@ -0,0 +1,199 @@
+import org.apache.hudi.table.HoodieTable;
+import org.apache.hudi.util.Lazy;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.apache.hudi.metadata.HoodieTableMetadata.SOLO_COMMIT_TIMESTAMP;
+import static org.apache.hudi.metadata.HoodieTableMetadataUtil.readRecordKeysFromBaseFiles;
+import static org.apache.hudi.metadata.MetadataPartitionType.RECORD_INDEX;
+
+public class RecordIndexer implements Indexer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RecordIndexer.class);
+  // Average size of a record saved within the record index.
+  // Record index has a fixed size schema. This has been calculated based on experiments with default settings
+  // for block size (1MB), compression (GZ) and disabling the hudi metadata fields.
+  public static final int RECORD_INDEX_AVERAGE_RECORD_SIZE = 48;
+  private final HoodieEngineContext engineContext;
+  private final HoodieWriteConfig dataTableWriteConfig;
+  private final HoodieTableMetaClient dataTableMetaClient;
+  private final HoodieTable table;
+
+  public RecordIndexer(HoodieEngineContext engineContext,
+                       HoodieWriteConfig dataTableWriteConfig,
+                       HoodieTableMetaClient dataTableMetaClient,
+                       HoodieTable table) {
+    this.engineContext = engineContext;
+    this.dataTableWriteConfig = dataTableWriteConfig;
+    this.dataTableMetaClient = dataTableMetaClient;
+    this.table = table;
+  }
+
+  @Override
+  public String getPartitionName() {
+    return RECORD_INDEX.getPartitionPath();
+  }
+
+  @Override
+  public Pair<Integer, HoodieData<HoodieRecord>> build(
+      List<HoodieTableMetadataUtil.DirectoryInfo> partitionInfoList,
+      Map<String, Map<String, Long>> partitionToFilesMap,
+      String createInstantTime,
+      Lazy<HoodieTableFileSystemView> fsView,
+      HoodieBackedTableMetadata metadata,
+      String instantTimeForPartition) throws IOException {
+    // TODO(yihua): could we unify the file listing across indexes?
+    // Collect the list of latest base files present in each partition
+    List<String> partitions = metadata.getAllPartitionPaths();
+    fsView.get().loadAllPartitions();
+    HoodieData<HoodieRecord> records = null;
+    if (dataTableMetaClient.getTableConfig().getTableType() == HoodieTableType.COPY_ON_WRITE) {
+      // for COW, we can only consider base files to initialize.
+      final List<Pair<String, HoodieBaseFile>> partitionBaseFilePairs = new ArrayList<>();
+      for (String partition : partitions) {
+        partitionBaseFilePairs.addAll(fsView.get().getLatestBaseFiles(partition)
+            .map(basefile -> Pair.of(partition, basefile)).collect(Collectors.toList()));
+      }
+
+      LOG.info("Initializing record index from " + partitionBaseFilePairs.size() + " base files in "
```

Comment:
Make sure to use the logger formatting `{}` instead of string concatenation throughout the PR

Reply:
Good catch.  Fixed.